### ISSUE TITLE
Error creating EC2 Instance in custom VPC in us-east-1b

### ISSUE THREAD

**Issue by timothyclifford** [docker/machine#584]
Description of problem:
Get the following error message when trying to create an Amazon EC2 host

`docker version`:
docker-machine version 0.1.0

`docker info`:

`uname -a`:
MINGW32_NT-6.1 LON03906 1.0.12(0.46/3/2) 2012-07-05 14:56 i686 unknown

Environment details (AWS, VirtualBox, physical, etc.):
AWS

How reproducible:
docker-machine create -d amazonec2 --amazonec2-access-key MY_ACCESS_KEY --amazonec2-secret-key MY_SECRET --amazonec2-vpc-id MY_VPC staging

Steps to Reproduce:
1. Run above command
2. Get error message

Actual Results:
ERRO[0001] Error creating machine: Error decoding error response: Error decoding error response: http: read on closed response body

Expected Results:
Docker host is created

Additional info:

**Comment by parshap** [comment]
I'm getting the same error on v0.1.0-rc5:
```
$ docker-machine -D create --driver amazonec2 --amazonec2-vpc-id $MY_VPC_ID docker-machine-test 
INFO[0000] Launching instance...                        
DEBU[0000] executing: /usr/bin/ssh-keygen ssh-keygen -t rsa -N  -f /Users/parshap/.docker/machines/docker-machine-test/id_rsa
DEBU[0000] creating key pair: docker-machine-test            
DEBU[0000] configuring security group in vpc-1172b274   
DEBU[0001] creating security group (docker-machine) in vpc-1172b274 
ERRO[0001] Error creating machine: Error decoding error response: Error decoding error response: http: read on closed response body 
WARN[0001] You will want to check the provider to make sure the machine and associated resources were properly removed. 
FATA[0001] Error creating machine                       
```

**Comment by ehazlett** [comment]
is your VPC configured to have public IPs?

**Comment by timothyclifford** [comment]
From what I can see yes everything is configured.

**Comment by ehazlett** [comment]
@timothyclifford @parshap sometimes it's a hiccup in the provider.  i've been creating ec2 instances today in us-east-1 without issue.  would you mind trying again?

**Comment by parshap** [comment]
@ehazlett: I'm getting a different error today.

```
$ docker-machine --debug create --driver amazonec2 --amazonec2-vpc-id $MY_VPC_ID docker-machine-test
INFO[0000] Launching instance...
DEBU[0000] executing: /usr/bin/ssh-keygen ssh-keygen -t rsa -N  -f /Users/parshap/.docker/machines/docker-machine-test/id_rsa

Generating public/private rsa key pair.
Your identification has been saved in /Users/parshap/.docker/machines/docker-machine-test/id_rsa.
Your public key has been saved in /Users/parshap/.docker/machines/docker-machine-test/id_rsa.pub.
The key fingerprint is:
fc:fa:bf:b9:99:de:7f:c9:0d:54:90:15:ba:65:09:3e parshap@bigbook.local
The key's randomart image is:
DEBU[0000] creating key pair: docker-machine-test
DEBU[0001] configuring security group in vpc-1172b274
ERRO[0007] Error creating machine: Problem with AWS API call: Non-200 API response: code=500 message=An internal error has occurred

WARN[0007] You will want to check the provider to make sure the machine and associated resources were properly removed.
FATA[0007] Error creating machine
```

Is there a way to print out the full AWS API request/response headers and body?

**Comment by parshap** [comment]
> is your VPC configured to have public IPs?

Yes. When I manually create an instance in the VPC I am able to SSH to it.

**Comment by ehazlett** [comment]
@parshap interesting -- the 500 leads me to believe that's on the Amazon side.  Do you happen to have another region configured in your environment or are you using us-east-1?  I believe @nathanleclaire had some amazon issues today too.

**Comment by jpetazzo** [comment]
I'm having the same issue today:

```
$ ./docker-machine --debug create -d amazonec2 jpetazzo3
INFO[0001] Launching instance...                        
DEBU[0001] executing: /usr/bin/ssh-keygen ssh-keygen -t rsa -N  -f /home/jpetazzo/.docker/machine/machines/jpetazzo3/id_rsa
 
Generating public/private rsa key pair.
Your identification has been saved in /home/jpetazzo/.docker/machine/machines/jpetazzo3/id_rsa.
Your public key has been saved in /home/jpetazzo/.docker/machine/machines/jpetazzo3/id_rsa.pub.
The key fingerprint is:
3d:13:ec:b8:cf:cc:8e:ed:19:a6:9c:e3:45:2a:bd:44 jpetazzo@tarrasque
The key's randomart image is:
+--[ RSA 2048]----+
|                 |
|         .       |
|          o      |
|         + .     |
|        E *      |
|       o + o     |
|      . = +      |
|       +.& o     |
|       .B+X      |
+-----------------+
DEBU[0001] creating key pair: jpetazzo3                 
DEBU[0001] configuring security group in vpc-a5f58ec0   
DEBU[0001] creating security group (docker-machine) in vpc-a5f58ec0 
ERRO[0002] Error creating machine: Error decoding error response: Error decoding error response: http: read on closed response body 
WARN[0002] You will want to check the provider to make sure the machine and associated resources were properly removed. 
FATA[0002] Error creating machine   
```

```
$ ./docker-machine -v
docker-machine version 0.1.0
```

Is there any way to get *moar debuggin'!* from the EC2 API?

**Comment by nathanleclaire** [comment]
> Is there any way to get moar debuggin'! from the EC2 API?

Dump the requests and responses, basically.  Unfortunately the API is not always super descriptive when things go wrong but it would be better than nothing.

I'd really like to have an audit log for all of the API requests and responses (where available), but implementing this cross-driver is tricky since some of them use mostly opaque 3rd party bindings.

**Comment by sthulb** [comment]
We could dump the response from the API. But ultimately, I have plans to replace most of the driver with the official SDK

**Comment by jeremyll** [comment]
__Subscribing__ - This issue makes machine unusable for me right now.

**Comment by ehazlett** [comment]
@jeremyll what version of machine are you using?

**Comment by jeremyll** [comment]
`0.1.0`

**Comment by ehazlett** [comment]
@jeremyll would you mind trying the latest RC?  It has several fixes for stability and retry.

**Comment by nerdalert** [comment]
@ehazlett in case you are still looking for data. The key pair does get created in EC2. I saw the same deserialization error. I didn't poke around in the code for a bug as I saw in #584 that the AWS Go API would likely replace the current. 

```
$ docker-machine -v
docker-machine version 0.2.0 (HEAD)

$ docker-machine -D create \
    --driver amazonec2 \
    --amazonec2-access-key $AWS_ACCESS_KEY_ID \
    --amazonec2-secret-key $AWS_SECRET_ACCESS_KEY \
    --amazonec2-vpc-id $AWS_VPC_ID \
    --amazonec2-subnet-id $AWS_SUBNET_EAST1 \
    test-instance

INFO[0000] Launching instance...
DEBU[0001] creating key pair: test-instance
DEBU[0001] configuring security group in vpc-8752d5e2
DEBU[0001] creating security group (docker-machine) in vpc-8752d5e2
ERRO[0001] Error creating machine: Error decoding error response: Error decoding error response: http: read on closed response body
WARN[0001] You will want to check the provider to make sure the machine and associated resources were properly removed.
FATA[0001] Error creating machine
```
Thanks

**Comment by nerdalert** [comment]
@ehazlett sorted me out. I had a mismatch in the VPC. Ty Evan!
- Pasting in output for future reference:

```
 $ docker-machine -v
docker-machine version 0.2.0 (HEAD)
$ docker-machine -D create \
> --driver amazonec2 \
> --amazonec2-access-key $AWS_ACCESS_KEY_ID \
> --amazonec2-secret-key $AWS_SECRET_ACCESS_KEY \
> --amazonec2-vpc-id $AWS_VPC_ID \
>  test-instance
INFO[0000] Launching instance...
DEBU[0001] creating key pair: test-instance
DEBU[0001] configuring security group in vpc-42dffc27
DEBU[0001] found existing security group (docker-machine) in vpc-42dffc27
DEBU[0001] configuring security group authorization for 0.0.0.0/0
DEBU[0001] launching instance in subnet subnet-209c260b
DEBU[0002] waiting for ip address to become available
DEBU[0021] Got the IP Address, it's "x.x.x.x"
DEBU[0021] created instance ID i-efc00f39, IP address x.x.x.x, Private IP address y.y.y.y
DEBU[0021] Settings tags for instance
DEBU[0021] Getting to WaitForSSH function...
DEBU[0177] generating server cert: /Users/brent/.docker/machine/machines/test-instance/server.pem ca-key=/Users/brent/.docker/machine/certs/ca.pem private-key=/Users/brent/.docker/machine/certs/ca-key.pem org=test-instance
INFO[0184] "test-instance" has been created and is now the active machine.
INFO[0184] To point your Docker client at it, run this in your shell: eval "$(docker-machine env test-instance)"
```

**Comment by ehazlett** [comment]
Thanks for the feedback!  I thought we checked if the specified VPC actually existed in the desired zone.  We should check that.

**Comment by timothyclifford** [comment]
Hi,

Just wanted to update this again as I've done a little more testing. Using v 0.3.0

Using a VPC in US West, I still get same error 

`Error decoding error response: Error decoding error response: http: read on closed response body`

Using a VPC in US East and I get the following:

```
C:\Program Files>docker-machine create -d amazonec2 --amazonec2-access-key KEY --amazonec2-secret-key SECRET --amazonec2-vpc-id ID --amazonec2-subnet-id ID --amazonec2-zone b docker-test

Launching instance...

To see how to connect Docker to this machine, run: docker-machine env docker-test
```

Looking in AWS, I can see the machine created successfully. So your suggestion about the problem being AWS region specific sounds spot on. 

I have triple checked that all the VPCs being used do exist and IDs were correct, very weird.

**Comment by ruffsl** [comment]
Having the same issues as @timothyclifford .
```shell
docker-machine -v
docker-machine version 0.3.0 (0a251fe)
```

**Comment by alvinr** [comment]
I'm trying to diagnose this issue with `docker-machine version 0.4.0-rc2 (c2b5f06)` - I'm on US West. I have this working on US East so I'm looking through to see if I can find any differences.

**Comment by alvinr** [comment]
Ok - here is the issue, its the way matching is done for Security Groups (in my case). I had set the following
```
export AWS_SECURITY_GROUP=sg-6d72d709
```

This is the Id rather than the name of the Security Group. If I use the Id, e.g.

```
export AWS_SECURITY_GROUP=alvin-demo-sg
```

Then the instance gets created just fine.

The code is question is https://github.com/docker/machine/blob/master/drivers/amazonec2/amazonec2.go#L608
```
	for _, grp := range groups {
		if grp.GroupName == groupName {
			log.Debugf("found existing security group (%s) in %s", groupName, d.VpcId)
			securityGroup = &grp
			break
		}
	}
```
The the EC2 API doc this check should be extended to do something like
```
if grp.GroupName == groupName  || grp. GroupId == groupName {
```

For more info see
http://docs.aws.amazon.com/cli/latest/reference/ec2/describe-security-groups.html

**Comment by nathanleclaire** [comment]
Nice debugging @alvinr !  Thanks!

**Comment by alvinr** [comment]
I think the above fix would resolve this issue 
https://github.com/docker/machine/issues/1568

**Comment by upgradingdave** [comment]
I ran into this same issue. When I specified security group by id `    --amazonec2-security-group sg-e46bc382 `, I got an error

     Launching instance...
     Error creating machine: Error decoding error response: Error decoding error response: http: read on closed response body
You will want to check the provider to make sure the machine and associated resources were properly removed.

But when I specified security group by name `--amazonec2-security-group my-group`, it worked. 

I'm running `docker-machine version 0.4.1 (e2c88d6)` on `OS X version 10.10.5`.
